File: 838-design-linked-list/design-linked-list.py
class LinkedList:
    def __init__(self, val=0):
        self.next = None
        self.val = val


class MyLinkedList:
    def __init__(self):
        self.left = LinkedList()  # Sentinel node for the head
        self.right = LinkedList()  # Sentinel node for the tail
        self.left.next = self.right
        self.size = 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        curr = self.left.next
        for _ in range(index):
            curr = curr.next
        return curr.val

    def addAtHead(self, val: int) -> None:
        node = LinkedList(val)
        node.next = self.left.next
        self.left.next = node
        self.size += 1

    def addAtTail(self, val: int) -> None:
        node = LinkedList(val)
        curr = self.left
        while curr.next != self.right:
            curr = curr.next
        node.next = self.right
        curr.next = node
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return
        if index == 0:
            self.addAtHead(val)
            return
        if index == self.size:
            self.addAtTail(val)
            return
        node = LinkedList(val)
        curr = self.left
        for _ in range(index):
            curr = curr.next
        node.next = curr.next
        curr.next = node
        self.size += 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        curr = self.left
        for _ in range(index):
            curr = curr.next
        curr.next = curr.next.next
        self.size -= 1

File: 838-design-linked-list/design-linked-list.java
class Node{
    Node next;
    int val;
    public Node(int val){
        this.next=null;
        this.val=val;
    }
    public Node(){
        this.next=null;
        this.val=0;
    }
}
class MyLinkedList {
    Node dummyHead;
    int size;

    public MyLinkedList() {
        this.dummyHead= new Node();
        this.size=0;
    }
    
    public int get(int index) {
        if(index<0||index>=size){
            return -1;
        }
        Node cur= dummyHead;
        for(int i = 0; i<=index;i++){
            cur=cur.next;
        }
        return cur.val;
    }
    
    public void addAtHead(int val) {
        Node node = new Node(val);
        node.next=dummyHead.next;
        dummyHead.next=node;
        size++;
    }
    
    public void addAtTail(int val) {
        Node node = new Node(val);
        Node cur= dummyHead;
        while(cur.next!=null){
            cur=cur.next;
        }
        cur.next=node;
        size++;
    }
    
    public void addAtIndex(int index, int val) {
        if(index<0||index>size){
            return;
        }
        if(index==0){
            addAtHead(val);
            return;
        }
        if(index==size){
            addAtTail(val);
            return;
        }
        Node node = new Node(val);
        Node cur= dummyHead;
        for(int i =0; i<index; i++){
            cur=cur.next;
        }
        node.next=cur.next;
        cur.next=node;
        size++;
    }
    
    public void deleteAtIndex(int index) {
        if(index<0||index>=size){
            return;
        }
        Node cur= dummyHead;
        for(int i =0 ;i<index; i++){
            cur=cur.next;
        }
        cur.next=cur.next.next;
        size--;
        
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */